### PR TITLE
Initial Halstead implementation for Java

### PR DESCRIPTION
**Pull request by dburriss**
- [x] Implement necessary functions
- [x] Passing test
- [x] Cleanup

Some references:

- [Java lang ref](https://docs.oracle.com/javase/specs/jls/se18/html/jls-3.html#jls-3.12)
- https://www.geeksforgeeks.org/software-engineering-halsteads-software-metrics/?msclkid=5e181114abef11ecbb03527e95a34828

### CODE REVIEW

**Comment by dburriss** [comment]
Test exmaple:

```java
public static void main(String args[]) {
  int a, b, c, avg;
  a = 5; b = 5; c = 5;
  avg = (a + b + c) / 3;
  MessageFormat.format(\"{0}\", avg);
}
```

I will get to this over the weekend and try to set my test up for expected numbers. 
@marco-c  who can explain to me the difference between `u_operators`, `operators`, `u_operands`, and `operands` in the Halstead tests?

**Comment by Luni-4** [comment]
> Test exmaple:
> 
> ```java
> public static void main(String args[]) {
>   int a, b, c, avg;
>   a = 5; b = 5; c = 5;
>   avg = (a + b + c) / 3;
>   MessageFormat.format(\"{0}\", avg);
> }
> ```
> 
> I will get to this over the weekend and try to set my test up for expected numbers. @marco-c who can explain to me the difference between `u_operators`, `operators`, `u_operands`, and `operands` in the Halstead tests?

@dburriss 
`u_operators` are all distinct operators present in the code
`u_operands` are all unique operands present in the code 
`operators` are all occurrences of `u_operators` in the code
`operands` are all occurrences of `u_operands` in the code.

You can find what are operators and operands in these tests https://github.com/mozilla/rust-code-analysis/blob/master/src/ops.rs#L312 for different programming languages

**Comment by dburriss** [comment]
@Luni-4 I have run into a weird issue.

When parsing the first line  `public static void main(String args[]){`

```
unknown: program 
unknown: local_variable_declaration 
unknown: modifiers 
unknown: public
unknown: static
operator: void_type
unknown: variable_declarator
operand: identifier
operator: ;
unknown: ERROR
unknown: formal_parameters
operator: (
unknown: formal_parameter
operator: type_identifier
operand: identifier
unknown: dimensions
operator: [
operator: ]
operator: )
unknown: block
```

Note the **operator: ;** and **unknown: ERROR** ... probably related. There is no **;** in the first line so this is throwing the count off.

The example is valid Java... I compiled it.

**Comment by calixteman** [comment]
It's very likely a matter of adding `public class A {`and `}` around your function.

**Comment by dburriss** [comment]
> It's very likely a matter of adding `public class A {`and `}` around your function.

Thanks @calixteman
That did the trick. It actually crossed my mind but I had never seen ERROR in the LoC parsing and most of those tests don't have class. Anyway, I learned something. And I am getting the  numbers I expect.

@Luni-4 I included a lot of comments and links, etc. in the code explaining what and why I am including different things. I didn't find definitive rules about what is included or excluded.

image

**Comment by dburriss** [comment]
@Luni-4 

I looked at *ops.rs* and added a test but having some problems.

Copying the existing tests quickly and filling in from the list above (or the code), I get the following error:

```
thread 'ops::tests::java_ops' panicked at 'assertion failed: `(left == right)`
  left: `["()", ")", "+", ",", ".", "/", ";", "=", "[]", "]", "integral_type", "method_invocation", "type_identifier", "void_type", "{}", "}"]`,
 right: `["()", "+", ",", "/", ";", "=", "String", "[]", "format", "int", "void", "{}"]`', src\ops.rs:280:9
```

A few questions questions:

1. Do you know why the list for Java is showing the grammar type instead of the value ie. 'integral_type` instead of `int` as it does for the other languages
2. Does the order of operators/operands matter in the assert?
3. Are these `println` supposed to be here? (hard to believe but they are not mine :) ) https://github.com/mozilla/rust-code-analysis/blob/master/src/ops.rs#L92

**Comment by dburriss** [comment]
@Luni-4 this is failing on the *ops.rs* file you mentioned. Not something I introduced. The linter has an opinion on something. I can make the change in this PR but I'm not sure how this got in in the first place. New rule maybe?

Log: https://community-tc.services.mozilla.com/tasks/WqxZrlJ2TIKZMbpDibX5Jg/runs/0/logs/public/logs/live_backing.log

**Comment by dburriss** [comment]
Shuffled things around a bit and moved ops check to #849
Hopefully, it is ok to merge those 2 separately.
@calixteman @Luni-4 ok to review with this scope?

halstead eg

File: src/getter.rs [hunk 1]
```diff
@@ -524,4 +524,42 @@ impl Getter for JavaCode {
...
+            | EQEQ | LTEQ | GTEQ | BANGEQ | AMPAMP | PIPEPIPE | PLUSPLUS | DASHDASH
+            | PLUS | DASH | STAR | SLASH | AMP | PIPE | CARET | PERCENT| LTLT | GTGT | GTGTGT
+            | PLUSEQ | DASHEQ | STAREQ | SLASHEQ | AMPEQ | PIPEEQ | CARETEQ | PERCENTEQ | LTLTEQ | GTGTEQ | GTGTGTEQ
+            // type identifier
+            | TypeIdentifier | IntegralType | FloatingPointType | BooleanType
+            => {
+                HalsteadType::Operator
+            },
+            // Operands: variables, constants, literals
+            Identifier | NullLiteral | ClassLiteral | StringLiteral | CharacterLiteral | HexIntegerLiteral | OctalIntegerLiteral | BinaryIntegerLiteral | DecimalIntegerLiteral | HexFloatingPointLiteral | DecimalFloatingPointLiteral  => {
```
Comment (Luni-4) [inline src/getter.rs:555]: Can we split this long line one more lines?

**Review by Luni-4 — CHANGES_REQUESTED** [review]
Sorry @dburriss for the delay, yeah, it's fine splitting up the two problems.

Just a final request and we can land this PR

File: src/getter.rs (hunk 1 above)
Comment (dburriss) [reply src/getter.rs]: I did find it weird the formatter didn't do that. Done 👍

**Review by Luni-4 — APPROVED** [review]
Thank you @dburriss! Fine for me!

**Pull request merged**